tianshou/data/batch.py

```diff
@@ -389,20 +201,13 @@ def __setitem__(self, index: Union[
             str, slice, int, np.integer, np.ndarray, List[int]],
             value: Any) -> None:
         """Assign value to self[index]."""
-        if isinstance(value, (list, tuple)):
-            value = np.asanyarray(value)
-        if isinstance(value, np.ndarray):
-            if not issubclass(value.dtype.type, (np.bool_, np.number)):
-                value = value.astype(np.object)
         if isinstance(index, str):
-            self.__dict__[index] = value
+            self.__setattr__(index, value)
             return
-        if not isinstance(value, (dict, Batch)):
-            if _is_batch_set(value):
-                value = Batch(value)
-            else:
-                raise TypeError("Batch does not supported value type "
-                                f"{type(value)} for item assignment.")
+        value = Batch(a=value).a
+        if isinstance(value, (np.ndarray, torch.Tensor)):
```

Comment:
it this really necessary to be that much restrictive ? I don't like the idea to add such a constraint for advanced users.

Reply:
I think setters of Batch should be consistent. For advanced users, they can directly manipulate ``self.__dict__``.

Reply:
> I don't like the idea to add such a constraint for advanced users.

This constraint is put on indexing with slice or int, not string. ``obj['a']=tensor`` still works, but ``obj[1] = tensor`` is really undefined, because ``obj[1]`` always returns a Batch object.